### PR TITLE
ENH: optimize.HessianUpdateStrategy: add `__matmul__`

### PR DESCRIPTION
**Pull request by dschmitz89**
#### Reference issue
This came up in https://github.com/scipy/scipy/issues/8792#issuecomment-1873466335

#### What does this implement/fix?
Allows to use the matmul operator @ for HessianUpdateStrategy.

#### Additional information
This is still missing a strict test that @ = .dot . 

CC @jeffrey-hokanson : any help reviewing would be much appreciated :)

### CODE REVIEW

**Comment by dschmitz89** [comment]
This is ready from my side. Last thing left: I am not sure how to document that `__matmul__` is implemened or if we want to do that.

File: scipy/optimize/_hessian_update_strategy.py [hunk 1]
```diff
@@ -382,7 +385,7 @@ def _update_implementation(self, delta_x, delta_grad):
             z = delta_x
         # Do some common operations
         wz = np.dot(w, z)
-        Mw = self.dot(w)
+        Mw = self@w
```
Comment (rgommers) [inline scipy/optimize/_hessian_update_strategy.py:388]: Note that `@` is an operator, so should have spaces around it (here and the 2 instances below).

**Review by rgommers — COMMENTED** [review]
Thanks @dschmitz89. This LGTM overall, only a couple of minor comments.

> Last thing left: I am not sure how to document that `__matmul__` is implemened or if we want to do that.

A basic example using `@` in the `Examples` section of the relevant classes seems like the way to do this to me.

**Comment by dschmitz89** [comment]
> Thanks @dschmitz89. This LGTM overall, only a couple of minor comments.
> 
> > Last thing left: I am not sure how to document that `__matmul__` is implemened or if we want to do that.
> 
> A basic example using `@` in the `Examples` section of the relevant classes seems like the way to do this to me.

Thanks for the review @rgommers . I looked into adding an example but as `HessianUpdateStrategy` is an abtract base class and neither `BFGS` or `SR1` already contained examples, I did not add a dedicated matmul example. Instead I added it in the sentence listing methods. Hope that is also fine for you.

**Review by rgommers — APPROVED** [review]
LGTM now, in it goes. Thanks @dschmitz89

**Pull request merged**